File: 2023/day03/main.lua
local function is_symbol(c)
    return c < '.' or c > '9' or c == '/'
end

local WIDTH = 140

local function make_adjacent_positions(i)
    return {
        i - 1,
        i - WIDTH - 1,
        i - WIDTH,
        i - WIDTH + 1,
        i + 1,
        i + WIDTH - 1,
        i + WIDTH,
        i + WIDTH + 1
    }
end

local schematic = {}
for line in io.lines("input") do

    for char in line:gmatch(".") do
        table.insert(schematic, char)
    end
end

local part_ref = {}
local part_positions = {}
local part_number = ""

for pos, character in ipairs(schematic) do
    if string.match(character, "%d") then
        part_number = part_number .. character
        table.insert(part_positions, pos)
    else
        for _, p in ipairs(part_positions) do
            part_ref[p] = tonumber(part_number)
        end
        part_positions = {}
        part_number = ""
    end
end

local part1 = 0
local part2 = 0
local parts_counted = {}

for i, char in ipairs(schematic) do
    if not is_symbol(char) then
        goto continue;
    end
    local first_part = 0
    for _, pos in ipairs(make_adjacent_positions(i)) do
        local part = part_ref[pos]
        if part ~= nil then
            -- N.B. No part numbers have more than one adjacent symbol
            if parts_counted[part] == nil then
                parts_counted[part] = true
                part1 = part1 + part
            end
            -- part 2
            if char == '*' then
                if first_part ~= 0 and part ~= first_part then
                    part2 = part2 + first_part * part
                end
                first_part = part
            end
        end
    end
    parts_counted = {}
    ::continue::
end

print("part1:", part1, "part2:", part2)

File: 2023/day03/main.rs
use std::collections::HashMap;
use std::collections::HashSet;

fn is_symbol(c: char) -> bool {
    c < '.' || c > '9' || c == '/'
}

const WIDTH: usize = 140;

fn make_adjacent_positions(i: usize) -> [usize; 8] {
    [
        i - 1,
        i - WIDTH - 1,
        i - WIDTH,
        i - WIDTH + 1,
        i + 1,
        i + WIDTH - 1,
        i + WIDTH,
        i + WIDTH + 1,
    ]
}

// Could we regex the lines of the file, get index of capture multiplied + line number * width to
// get pos
fn main() {
    let schematic = include_str!("input").lines().flat_map(|line| line.chars());

    let mut part_ref = HashMap::<usize, i32>::new();
    let mut part_positions = Vec::<usize>::new();
    let mut part_number = String::new();

    schematic.clone().enumerate().for_each(|(pos, character)| {
        if character.is_digit(10) {
            part_number.push(character);
            part_positions.push(pos);
            return;
        }
        for p in part_positions.iter() {
            part_ref.insert(*p, part_number.parse::<i32>().unwrap());
        }
        part_positions.clear();
        part_number.clear();
    });

    let mut part1 = 0;
    let mut part2 = 0;
    let mut parts_counted = HashSet::<i32>::new();
    for (i, char) in schematic.enumerate() {
        // part 1
        if !is_symbol(char) {
            continue;
        }
        let mut first_part = 0;
        for pos in make_adjacent_positions(i).iter() {
            if part_ref.contains_key(pos) {
                let part = part_ref[pos];
                // N.B. No part numbers have more than one adjacent symbol
                if parts_counted.insert(part) {
                    part1 += part;
                }
                // part 2
                if char == '*' {
                    if first_part != 0 && part != first_part {
                        part2 += first_part * part;
                    }
                    first_part = part;
                }
            }
        }
        parts_counted.clear();
    }

    println!("part1: {}, part2: {}", part1, part2);
}
